ext/libclementine-tagreader/tagreader.cpp

```diff
@@ -130,6 +130,51 @@ const char* kASF_OriginalDate_ID = "WM/OriginalReleaseTime";
 const char* kASF_OriginalYear_ID = "WM/OriginalReleaseYear";
 }  // namespace
 
+void TagReader::RemoveExtension(std::string* s) const {
+  if (s->empty()) return;
+  const int i = s->find_last_of('.');
+  if (i == std::string::npos) return;
+  s->assign(s->substr(0, i));
+}
+
+void TagReader::ChangeUnderscores(std::string* s) const {
+  for (int i = 0; i < s->size(); i++) {
+    if (s->at(i) == '_') s->at(i) = ' ';
+  }
+}
+
+std::string& TagReader::ltrim(std::string& str, const std::string& chars) const {
+    str.erase(0, str.find_first_not_of(chars));
+    return str;
+}
+
+std::string& TagReader::rtrim(std::string& str, const std::string& chars) const {
+    str.erase(str.find_last_not_of(chars) + 1);
+    return str;
+}
+
+std::string& TagReader::trim(std::string& str, const std::string& chars) const {
+    return ltrim(rtrim(str, chars), chars);
+}
+
+void TagReader::GuessArtistAndTitle(pb::tagreader::SongMetadata *song) const {
+  const std::string bn = song->basefilename();
+  std::string *artist = song->mutable_artist();
```

Comment:
Calling `mutable` means that `has_artist()` will no longer return `false` even if you leave it empty.